source/common/network/udp_listener_impl.cc

```diff
@@ -94,56 +71,75 @@ void UdpListenerImpl::onSocketEvent(short flags) {
 
 void UdpListenerImpl::handleReadCallback() {
   ENVOY_UDP_LOG(trace, "handleReadCallback");
-  sockaddr_storage addr;
-  socklen_t addr_len = 0;
-
+  // TODO(danzh) make this variable configurable to support jumbo frames.
+  const uint64_t read_buffer_length = MAX_UDP_PACKET_SIZE;
   do {
-    ReceiveResult recv_result = doRecvFrom(addr, addr_len);
-    if ((recv_result.result_.rc_ < 0)) {
-      if (recv_result.result_.errno_ != EAGAIN) {
-        ENVOY_UDP_LOG(error, "recvfrom result {}", recv_result.result_.errno_);
+    Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+    Buffer::RawSlice slice;
+    const uint64_t num_slices = buffer->reserve(read_buffer_length, &slice, 1);
+    ASSERT(num_slices == 1);
+
+    IoHandle::RecvMsgOutput output(&packets_dropped_);
+    uint32_t old_packets_dropped = packets_dropped_;
+    MonotonicTime receive_time = time_source_.monotonicTime();
+    Api::IoCallUint64Result result = socket_.ioHandle().recvmsg(
+        &slice, num_slices, socket_.localAddress()->ip()->port(), output);
+
+    if (!result.ok()) {
+      // No more to read or encountered a system error.
+      if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+        ENVOY_UDP_LOG(error, "recvfrom result {}: {}",
+                      static_cast<int>(result.err_->getErrorCode()),
+                      result.err_->getErrorDetails());
         cb_.onReceiveError(UdpListenerCallbacks::ErrorCode::SyscallError,
-                           recv_result.result_.errno_);
+                           result.err_->getErrorCode());
       }
+      // Stop reading.
       return;
     }
 
-    if (recv_result.result_.rc_ == 0) {
+    if (result.rc_ == 0) {
       // TODO(conqerAtapple): Is zero length packet interesting?
-      return;
+      ENVOY_UDP_LOG(warn, "received 0-length packet");
     }
 
-    Address::InstanceConstSharedPtr local_address = socket_.localAddress();
+    if (output.local_address_ == nullptr) {
+      ENVOY_UDP_LOG(error, "fail to get local address from IP header");
+    }
 
-    RELEASE_ASSERT(
-        addr_len > 0,
-        fmt::format(
-            "Unable to get remote address for fd: {}, local address: {}. address length is 0 ",
-            socket_.ioHandle().fd(), local_address->asString()));
+    if (packets_dropped_ != old_packets_dropped) {
+      // The kernel tracks SO_RXQ_OVFL as a uint32 which can overflow to a smaller
+      // value. So as long as this count differs from previously recorded value,
+      // more packets are dropped by kernel.
+      uint32_t delta = (packets_dropped_ > old_packets_dropped)
+                           ? (packets_dropped_ - old_packets_dropped)
+                           : (packets_dropped_ +
+                              (std::numeric_limits<uint32_t>::max() - old_packets_dropped) + 1);
+      // TODO(danzh) add stats for this.
+      ENVOY_UDP_LOG(warn, "Kernel dropped {} more packets. Consider increase receive buffer size.",
```

Comment:
debug level

Reply:
Unlike the empty UDP packet warning, this warning shouldn't cause log spam, because it logs only when kernel drops more packets. If this message floods the log, the connection must be experiencing extreme high packet loss. At that point, other alerts probably already fired.

That's saying, this is unlikely to happen outside of production, except for stress tests. If we don't want it to show up in production, it's better to remove the log entirely.

Reply:
If the machine is overloaded, this can definitely happen, and log spam will only make it worse. This definitely needs a stat that we can alarm on, but for now since this is in the data path can we change it to debug?

Reply:
done